Flaky rpython memory tests

These two failures:

https://github.com/pypy/pypy/actions/runs/7533803201/job/20507000160
https://github.com/pypy/pypy/actions/runs/7533317545/job/20505615167

and this success:

https://github.com/pypy/pypy/actions/runs/7538948557/job/20520450692

are with the same code (I removed a line of whitespace to retrigger the test). A similar suspected flake is here: https://github.com/pypy/pypy/actions/runs/7466932227/job/20319376709

The failures were:

```python
____________________ TestIncrementalMiniMarkGC.test_weakref ____________________

self = <rpython.memory.test.test_transformed_gc.TestIncrementalMiniMarkGC object at 0x0000000064eaf2b8>

    def test_weakref(self):
        run = self.runner("weakref")
>       res = run([])

rpython/memory/test/test_transformed_gc.py:523: 
<... snip .../>
self = <C object Struct timeval { c_tv_sec, c_tv_usec } at 0x6805d60>
other = <struct header { tid=<CombinedSymbolic GroupMemberOffset(* <rpython.rtyper.lltypesystem.llgroup.group object at 0x00000000070fed08>, 3)|279172874240> }>

    def __eq__(self, other):
        if isinstance(other, _llgcopaque):
            addressof_other = other.intval
        else:
            if not isinstance(other, lltype._parentable):
                return False
            if self._storage is None or other._storage is None:
>               raise RuntimeError("pointer comparison with a freed structure")
E               RuntimeError: pointer comparison with a freed structure
```

and 

```python
_______________ TestIncrementalMiniMarkGC.test_nursery_hash_base _______________

self = <rpython.memory.test.test_transformed_gc.TestIncrementalMiniMarkGC object at 0x0000000063bd8c98>

    def test_nursery_hash_base(self):
        res = self.runner('nursery_hash_base')
>       assert res([]) >= 195

rpython/memory/test/test_transformed_gc.py:1373: 
<... snip .../>
self = <C object Struct timeval { c_tv_sec, c_tv_usec } at 0xc888c80>
other = <struct header { tid=-42 }>

    def __eq__(self, other):
        if isinstance(other, _llgcopaque):
            addressof_other = other.intval
        else:
            if not isinstance(other, lltype._parentable):
                return False
            if self._storage is None or other._storage is None:
>               raise RuntimeError("pointer comparison with a freed structure")
E               RuntimeError: pointer comparison with a freed structure
```

the suspected one was also a `RuntimeError: pointer comparison with a freed structure` during `__eq__`.

I'm unsure exactly how to investigate this (especially as I can't actually get it to fail locally and subsequent runs in the pipeline pass). Possibly somebody has ideas about what could be causing `_storage` to be set to `None` when it shouldn't be? Or is a call to `__eq__` a symptom of something else failing (e.g. having to resolve a hash collision..?)

--- a/rpython/memory/test/test_transformed_gc.py
+++ b/rpython/memory/test/test_transformed_gc.py
@@ -52,6 +52,12 @@ class GCTest(object):
     taggedpointers = False
     gchooks = None
 
+    def setup(self):
+        # This snippet is an attempt to clean the WeakKeyDict
+        # rpython.memory.gcheader.header2obj between tests
+        import gc
+        gc.collect()
+
     def setup_class(cls):
         cls.marker = lltype.malloc(rffi.CArray(lltype.Signed), 1,
                                    flavor='raw', zero=True)

